abi.html

```diff
@@ -4107,73 +4106,156 @@ <h5><a href="#dso-dtor-motivation"> 3.3.6.1 Motivation </a></h5>
+reverse order of registration.  Calls will be made from the
+registering thread.
 
 <p>
-<li> Termination:
+If the process is exited by a thread, it is unspecified whether
+thread-local termination functions registered for different
+threads are called.  This applies even if another thread has
+terminated but has not yet finished calling all of its
+termination functions.
+
 <p>
-When linking any DSO containing a call to <code>__cxa_atexit</code>,
+If the unloading of a DSO is required other than during process
+exit (e.g. by calling <code>dlclose</code>), and there are any
+any thread-local termination functions associated with that DSO,
+the behavior is undefined.  Some implementations choose to support
+this by delaying the unloading of the DSO until all such functions
+have been called on all threads.  Implementations that wish to
+unload the DSO immediately are encouraged to at least call
+thread-local termination functions registered for the current
+thread.
+
+<p>
+<code>__cxa_thread_atexit</code> returns zero if registration
+is successful, nonzero on failure.
+
+<a name="atexit">
+<h5><a href="#atexit">3.3.6.5 User calls to <code>atexit</code></a></h5>
+
+<p>
+The C++ standard requires calls to exit functions registered with
+<code>atexit</code> to occur in reverse order of registration,
+appropriately interordered with the destruction of global objects
+according to the completion order of their initialization; see
+<span class="cxxref">[basic.start.term]</span>.  It is expected
+that implementations supporting C++ will integrate support for
+<code>__cxa_atexit</code> into the C library in order to achieve
+these semantics.
+
+<p>
+There is no analogous thread-specific <code>atexit</code> API
+in the C or C++ standards, or in POSIX threads.  If there were,
+it would be expected to be appropriately interordered with the
+destruction of thread-local objects.  It is currently unclear
+what the appropriate interordering rules are for the destruction
+of thread-local storage such as C11's <code>tss_create</code> or
+POSIX's <code>pthread_key_create</code>; these APIs do not
+specify any particular destruction order for multiple keys.
+
+<p>
+The C++ standard does not provide a user interface corresponding
+to <code>__cxa_atexit</code> or <code>__cxa_thread_atexit</code>.
+There is no way to register an <code>atexit</code> function
+with a parameter or a home DSO.
+
+<a name="termination">
+<h5><a href="#termination">3.3.6.6 Termination</a></h5>
+
+<p>
+When linking any DSO containing a call to <code>__cxa_atexit</code>
+or <code>__cxa_thread_atexit</code>,
 the linker should define a hidden symbol <code>__dso_handle</code>,
 with a value which is an address in one of the object's segments.
-(It does not matter what address,
-as long as they are different in different DSOs.)
-It should also include a call to the following function in the FINI
+(It does not matter what address, as long as they are different in
```

Comment:
In the GNU implementation, it must be an address within the object defining that hidden symbol (not an absolute address). I suggest to drop the remark in parentheses, it's confusing.

Reply:
I agree that the definition of `__dso_handle` is arguably outside of the ABI's scope; I'll think about how to reword this.